src/parallel/parallel.cpp

```diff
@@ -1,31 +1,126 @@
 #include "parallel.hpp"
 
+constexpr unsigned indices_per_segment = 16;
+
+////////////////////////////////////////////////////////////////////////////////
+/// SEGMENT LOCK
+////////////////////////////////////////////////////////////////////////////////
+
+size_t
+SegmentLock::get_counter()
+{
+    return this->counter_;
+}
+
+bool
+SegmentLock::is_locked()
+{
+    bool locked_by_me = this->mutex_.try_lock();
+    if (locked_by_me) {
+        this->mutex_.unlock();
+        return false;
+    }
+    return true;
+}
+
+void
+SegmentLock::lock()
+{
+    this->mutex_.lock();
+    ++this->counter_;
+}
+
+void
+SegmentLock::unlock()
+{
+    this->mutex_.unlock();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// BUCKET
+////////////////////////////////////////////////////////////////////////////////
+
+void
+ParallelBucket::lock()
+{
+    this->mutex.lock();
+}
+
+void
+ParallelBucket::unlock()
+{
+    this->mutex.unlock();
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 /// THREAD MANAGER
 ////////////////////////////////////////////////////////////////////////////////
 
 void
-ThreadManager::lock(size_t next_index)
+ThreadManager::lock(size_t index)
 {
-    // TODO
+    size_t segment_index = this->get_segment_index(index);
+    this->hash_table_->segment_locks_[segment_index].lock();
+    this->locked_segments_.push_back(segment_index);
 }
 
 void
+ThreadManager::release_all_locks()
+{
+    for (auto i : this->locked_segments_) {
+        this->hash_table_->segment_locks_[i].unlock();
+    }
+    this->locked_segments_.resize(0);
+}
```

Comment:
Can we maybe use a `std::vector<std::scoped_lock>` to handle this? Then we can do:
Lock:
```c++
this->locked_segments_.push_back(std::scoped_lock(segment_lock));
```
Release all:
```c++
this->locked_segments_.resize(0);
```

Where RAII handles calling unlock in release all.

I think this might be better for exception safety.